### ISSUE TITLE
The speed on the android devices is very slow

### ISSUE THREAD

**Issue by Bequiet19** [jurialmunkey/plugin.video.themoviedb.helper#617]
First I would like to thank you for this great addon. I've been using it on my PC for a long time. However, it doesn't work properly on Android devices. Opening any list is sometimes a big torture. Waiting sometimes 20 seconds to open the list. Especially if I turn on to uses fanart images.................. I have tried on several android devices with different configurations and the same thing. Let's say some other addons open lists quickly, especially elementum that opens almost instantly. I know that the addons in python are not speed champions, but is there any chance that the opening lists on Android devices will speed up a bit. Thanks in advance.

**Comment by jurialmunkey** [comment]
Don't turn on fanarttv or extra details.

**Comment by jurialmunkey** [comment]
Specifically these options will slow down loading for new items that haven't been cached yet (from most expensive to least expensive)

1. `API Keys > Use fanart.tv (all options)`
2. `General > Get additional details and artwork from TMDb`
3. `Accounts > Get watched indicators from Trakt`
4. `General > Get additional details from local Kodi library database`

FanartTV is a slow website, and if you don't have a VIP key it is even slower (sometimes a request can take upwards of ~800ms).

TMDb is faster. However, even on a fast machine (i7+SSD+Cable), the python requests module still logs a ~400ms return for a single request to the TMDb API. So just for one list, TMDb details adds 20 requests and since cheap android sticks suck at multithreading that's going to be at least 8 seconds where TMDbHelper is just waiting for the data from the website. 

Add fanart.tv on top of that with a non-VIP account and it's pretty easy to see how you hit 20 seconds for a list of 20 items on a cheap android stick.

I can't make the python requests module go faster or your cheap Android stick be better at multithreading... And it's simply not possible to compete against binary blob executables compiled for specific platforms like Elementum does.

**Comment by Bequiet19** [comment]
@jurialmunkey When I said I’ve tried on multiple android devices I didn’t just mean cheap android sticks. I tried both on a tablet and a mobile phone with quite serious configurations. Simply opening the list is far slower than on a PC. But ok if you say nothing can be done. I'm sorry about that since I thought I'd use tmdb helper on android. But at speeds like this, I'll get around it. It definitely stays on the pc.

**Comment by jurialmunkey** [comment]
Like I said, the number one issue is the speed of requests. Just a pure request to TMDb website via python requests module is ~400ms round trip at the best of times. If we take a realistic average of 500ms per request then that's 8 seconds on a 20 item list just for TMDb data.

On a PC, the cpus support multithreading so the requests can run in a parallel which significantly speeds things up because we aren't wasting time idling waiting for the data to return. However, on a mobile device, the cpu is geared towards battery life and typically doesn't support multithreading, so the requests are forced to happen in series one at a time instead.

It's the type of cpu in use not the Android platform per se. It doesn't matter how powerful the cpu is, if it can't process parallel threads then it's going to have to sit and wait for the server to respond with the data for each call before it moves onto the next instruction. Processing power can't make a server provide a request faster. 

There are ways to handle these types of scenarios better on mobile devices but not with standard Python modules.

**Comment by Bequiet19** [comment]
@jurialmunkey I understand everything. It's just a shame it's like that. We all know that android is being used more and more for kodi. But at speeds like this, it really seems useless.

**Comment by Bequiet19** [comment]
@jurialmunkey My dear friend, I found an addon (Fen) based on python that is lightning fast when opening all lists. With all included metadata and fanart images it opens up almost everything instantly (on pc and android devices). So far I've tried many similar addons but none are as fast to open lists as this one (not even the elementum). I don't know how this was achieved. I'd like to see if it's possible to do it the same with our beloved tmdb helper. Thank you in advance.
https://github.com/Tikipeter/repository.tikipeter

**Comment by Bequiet19** [comment]
@jurialmunkey Did you maybe check out this addon I sent you? Is there a chance to get it at this speed in tmdb helper? I think this kind of acceleration would significantly improve such a good addon as tmdb helper.

**Comment by jurialmunkey** [comment]
That add-on doesn't get anywhere close to the same amount of metadata that TMDbHelper does.

I gave it a quick test and it didn't seem that much faster than TMDbHelper to me. It is definitely quicker but not massively so.

~Also, from a quick look at the code, it only does Fanart TV lookups IF the item do not have poster/fanart available on TMDb. Since TMDb nearly always has artwork, that means Fen is basically never looking up Fanart TV. I would expect that if you disable FanartTV in TMDbH you will probably get comparable performance~ Actually even with fanarttv forced it is quite fast on that side of things which I have no idea how it is managing to do that.

Without knowing what exactly is causing slow downs on your system though, there is no way to know what exactly it is doing differently that would result in a speed gain for you. I can't just rewrite the entire plugin - there is is a massive amount of code involved.

**Comment by ambulancePilot** [comment]
Sorry to comment here. If I disable the fanart settings you mentioned above, what exactly will I be missing? Does clearart come from fanart.tv or TMDB?

**Comment by jurialmunkey** [comment]
> Sorry to comment here. If I disable the fanart settings you mentioned above, what exactly will I be missing? Does clearart come from fanart.tv or TMDB?

TMDb provides posters, fanart, landscape, clearlogo 
FanartTV provides posters, fanart, landscape, clearlogo, banners, characterart, clearart, discart.

You should only really need to enable fanarttv if you need clearart. If you use a skin that utilises the service monitor (e.g. Arctic Horizon) then you can get away with only enabling fanarttv for the service monitor.

By only enabled fanarttv for service and disabling it for plugin/widget paths, you'll get clearart "on the fly" for the focused item but the plugin directories will load faster because they're not being forced to lookup the fanarttv artwork of every item first.

Also the service monitor caches the details it looks up, so next time plugin directory loads that particular item the cached artwork details will automatically be attached directly to the item.

**Comment by jurialmunkey** [comment]
@ambulancePilot @Bequiet19 - In latest git version (v4.5.8) I've add a timer logging function to allow you to benchmark what is causing slow downs. You can enable it `TMDbHelper Settings > Expert > Log timer reports for directories` (it's just below the "Delete Cache" function).

That will output into kodi.log a report whenever you open any plugin directory. Note that it will add overhead to do the logging so remember to turn it off when you're done!

It will output something like the following in the log:
```
INFO <general>: [plugin.video.themoviedb.helper]
   DIRECTORY TIMER REPORT
   info=trakt_userlist&list_name=Rotten%20Tomatoes%3a%20Best%20of%202018&list_slug=rotten-tomatoes-best-of-2018&plugin_category=Rotten%20Tomatoes%3a%20Best%20of%202018&user_slug=lish408
   get_list       :   1.242 sec
    - kodi_db     :   0.714 sec
   add_items      :   3.749 sec
    - item_tmdb   :   0.708 sec (Average) |  15
    - item_ftv    :   1.805 sec (Average) |  19
    - item_kodi   :   None
    - item_trakt  :   0.393 sec (Average) |   1
    - item_build  :   0.023 sec (Average) |  20
   total          :   4.990 sec
   ------------------------------
   
   item_tmdb:
   0.062  0.625  0.596  0.681  0.677  0.712  0.672  0.736  0.748  0.755  0.767  0.775  0.851  0.890  1.080 
   
   item_ftv:
   0.142  0.145  0.089  0.381  1.026  1.394  1.263  2.322  2.346  2.143  2.438  2.437  2.663  2.570  2.600  2.292  2.586  2.760  2.700 
```

That output is on a Trakt list I've never opened with every setting enabled, so overall about the slowest type of scenario possible (other than first sync of inprogress tv which is particularly slow).

**Understanding the Output**

**get_items**
Time it takes to get the base list from the API, do some basic configuration, and also grab Kodi's library. You can see from the `kodi_db` line that I could save 0.714 of that 1.242 by disabling the Kodi library lookup setting.

Retrieving my library is slow here because I'm using my large dummy test library with 1000+ movies in it. My real library of approx 250 movies is much faster (around 0.300s max)

**add_items**
Time it takes to retrieve all the extra TMDb, FanartTV, Trakt, and Kodi details for all the items. This is the total time of all the requests. Each part of the request is itemised underneath. 

You can see that the most expensive is FanartTV. The value is the average time of all threads which should give a fairly decent picture of what the typical cost is. The individual thread timings are provided below the table in a list

**total**
That's how long the entire process took from requesting the plugin directory to displaying the items on screen.

**Comment by Bequiet19** [comment]
@jurialmunkey 
These are my average results. I took a list I didn’t open to have realistic results.The latest version of tmdb helper. There is little improvement. The biggest delay is with main movie fanart image. When I switch to the next movie everything appears almost instant (clearlogo, ratings, studio logo, all metadata... ) except the fanart movie image which is about 3 seconds late. This is from an android device.
```
DIRECTORY TIMER REPORT
                                                   info=trakt_userlist&list_name=Rotten%20Tomatoes%3a%20Best%20of%202018&list_slug=rotten-tomatoes-best-of-2018&plugin_category=Rotten%20Tomatoes%3a%20Best%20of%202018&user_slug=lish408
                                                   ------------------------------
                                                   get_list       :   1.099 sec
                                                    - kodi_db     :   None
                                                   add_items      :   8.453 sec
                                                    - item_tmdb   :   1.859 sec avg |   2.704 sec max |  21
                                                    - item_ftv    :   3.788 sec avg |   5.444 sec max |  20
                                                    - item_kodi   :   None
                                                    - item_trakt  :   None
                                                    - item_build  :   0.058 sec avg |   1.219 sec all |  21
                                                   ------------------------------
                                                   Total          :   9.552 sec
                                                   
                                                   item_tmdb:
                                                   0.243  0.051  1.288  1.789  1.915  1.915  1.996  2.048  2.094  2.100  1.619  1.943  1.977  1.974  1.759  2.449  2.494  2.059  2.262  2.357  2.704 
                                                   
                                                   item_ftv:
                                                   0.182  2.772  3.060  2.548  2.535  3.317  3.128  2.941  3.802  3.960  4.491  4.012  4.517  5.444  4.905  5.136  4.970  4.672  4.853  4.509
```

**Comment by jurialmunkey** [comment]
> When I switch to the next movie everything appears almost instant (clearlogo, ratings, studio logo, all metadata... ) except the fanart movie image which is about 3 seconds late

How long images take to load has nothing to do with Tmdbhelper. Once the items appear on screen, tmdbhelper is finished. Tmdbhelper just provides the image url to kodi, it doesn't download the image itself. It is kodi which downloads the image and displays it on screen.

Tmdbhelper provides ALL the metadata for the items at the same time. If there is a delay with images displaying on screen but the items are there then that's to do with kodi, your internet connection, and/or the website the image is from. If this is your main issue then it's not tmdbhelper which is slow.

**Comment by Bequiet19** [comment]
@jurialmunkey
I understand, but this kind of thing has never happened before. I compared this with version 4.4.0 from kodi repo. In version 4.4.0 list opening is slower but when I switch from movie to movie there is not deley. Opening the list itself is faster in last version, but when it enters the list, the appearance of the fanart image is delayed, for what reason I don't know. Also when I switch from one movie to another the same thing happens. Again I say the opening of the list has improved but something has happened with this. If I need to, I can record a video to see what it's all about.

**Comment by jurialmunkey** [comment]
> @jurialmunkey I understand, but this kind of thing has never happened before. I compared this with version 4.4.0 from kodi repo. In version 4.4.0 list opening is slower but when I switch from movie to movie there is not deley. Opening the list itself is faster in last version, but when it enters the list, the appearance of the fanart image is delayed, for what reason I don't know. Also when I switch from one movie to another the same thing happens. Again I say the opening of the list has improved but something has happened with this. If I need to, I can record a video to see what it's all about.

A video would be helpful.

Are you sure that these fanart images that are taking longer to load are simply not new images whereas the previous version was an already cached image? Once Kodi has downloaded an image it adds it to the texture cache - so an old image will show as soon as the url is there; but for a new image there will always be a delay because it needs to be downloaded first - you'll see this even in the library when you add a new item.

The only reason I can think of for a delay in image loading caused by TMDbHelper would be if the skin is getting the fanart from the service monitor rather than the item itself, which is a really odd approach and I don't know why a skin would do that. And anyway, it drops all the artwork properties at the same time, so the fanart url should be available at roughly the same time as the clearart etc.

**Comment by Bequiet19** [comment]
@jurialmunkey
To get realistic results I deleted all thumnails and all caches. The first is tmdb helper version 4.4.0. As you can see, all the images are displayed at the same time without delay.
https://files.fm/f/87kzg9t2g
The second is the latest version of tmdb helper. As you can see there is a delay when switching movies.
https://files.fm/f/r4c6tgdnw
Both lists are the same. Tested under the same conditions. Sorry for the bad videos and the way I pinned the files but github doesn't support large files.

**Comment by jurialmunkey** [comment]
@Bequiet19 

Did you delete `Textures13.db` from userdata? That's the one that counts.

Also, what skin is that? I'd like to look at the code. It looks like it is getting fanart from the service monitor which is a really odd approach. Still, I just can't see any reason why any changes since 4.4.0 would result in the artwork properties suddenly being filled slower than ratings (but that looks like what is happening) - I almost always get both threads finishing at the same time and if anything the ratings lag behind because that thread starts later and has more work to do.

**Comment by Bequiet19** [comment]
@jurialmunkey
Of course. During both tests I deleted Textures13.db. This is titan bingie mode 2.0.3.
https://github.com/AchillesPunks/repository.titan.bingie.mod

**Comment by Bequiet19** [comment]
@jurialmunkey
It's not up to the skin. I just tried aeon nox the same thing happens. A poster appeared and then after two or three seconds a fanart image.

**Comment by jurialmunkey** [comment]
> @jurialmunkey It's not up to the skin. I just tried aeon nox the same thing happens. A poster appeared and then after two or three seconds a fanart image.

If it's happening in Aeon Nox then it can't be TMDbHelper because Aeon Nox doesn't use the service monitor. It is impossible for a plugin to modify the properties of a listitem after it is created - so if the list of items is on screen then the artwork property is already there. 

If you are seeing a delay in Aeon Nox then that's entirely to do with how long it is taking Kodi to download the image. Most likely what is happening is that in one version of TMDbHelper you have the image coming from TMDb and in the other you have it coming from FanartTV, and one is slower than the other (or higher resolution, larger file size etc.).

You can inspect the properties by creating an xml file like the following and dropping it in your skin's `1080i` folder (or it might be `16x9` or `xml` depending on the skin - its the one where the rest of the skin xml files are).
[Custom_Overlay_1189.zip](https://github.com/jurialmunkey/plugin.video.themoviedb.helper/files/7985956/Custom_Overlay_1189.zip)

This will show you at the top of the screen what the url for the fanart is.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<window type="dialog" id="1189">
    <zorder>2</zorder>
    <visible>true</visible>
    <controls>
        <control type="label">
            <height>40</height>
            <textcolor>magenta</textcolor>
            <label>$INFO[ListItem.Art(fanart)]</label>
        </control>
    </controls>
</window>
```

**Comment by jurialmunkey** [comment]
@Bequiet19 - It'll be because a while back FanartTV changed how they handled "no language" items so that posters without text got the language `"00"` rather than `"en"`. But for some reason they made a _different_ (and undocumented) change for fanart where it went from default to `"en"` to now `""` (instead of `"00"` like the change with posters). 

No idea why the inconsistency but the side effect of me adding a language fix meant that fanart started coming solely from TMDb. TMDb recently started providing 4k fanart which is a much higher quality than most of the 1080p art on fanarttv, hence the longer download time. 

Anyway, 4.5.14 will force fanart from fanarttv if the option is enabled so that the behaviour is the same as it was in 4.4.0 prior to the API change.

**Comment by Bequiet19** [comment]
@jurialmunkey
The situation is a little better. The same thing still happens with some movies. I don't know why this is happening. Maybe because they are, as you say, bigger pictures, I don't know.

**Comment by matke-84** [comment]
@jurialmunkey
Jurial my friend, you did a very good job with this update. You remember that about a year ago we talked about this that you need to speed up tmdb helper on android. Now we are very close to tmdb helper being the ideal addon for android. Maybe if we could get two or three seconds more at speed it would be perfect. In any case, go ahead. Keep developing this great addon. All the best my friend. 😉

**Comment by matke-84** [comment]
@jurialmunkey
Hmmmm. There is indeed some delay with the fanart. Now I looked at the size of the downloaded fanart files. All are in full hd resolution and the size is at most about 500kb. So I'm not sure it's up to file size. Again, on the other hand, it is not clear to me since it downloads everything from fanart.tv, why it downloads one image quickly and needs more time for another. Really very strange situation. 🤔

**Comment by drinfernoo** [comment]
Fanary.tv is notorious for being rather slow to respond to requests, as well.

**Comment by matke-84** [comment]
@drinfernoo
I agree. But as far as I saw in the video posted by @Bequiet19 I see that with the old version it was not a problem. That's what's interesting.

**Comment by jurialmunkey** [comment]
> @jurialmunkey Hmmmm. There is indeed some delay with the fanart. Now I looked at the size of the downloaded fanart files. All are in full hd resolution and the size is at most about 500kb. So I'm not sure it's up to file size. 

That's incorrect. If you look in the videos you can clearly see the Artwork for Avengers End Game is different in each so not a fair comparison at all. Also files in Thumbnails folder are similar sizes because Kodi texture cache has downsampled them.

In the "slow" video the artwork is coming from TMDb:
https://www.themoviedb.org/t/p/original/7RyHsO4yDXtBv1zUU3mTpHeQ0d5.jpg
That artwork is *1617kb* and is *2160p*.  With default Kodi advancedsettings, Kodi's texture cache uses a bicubic filter to downsample fanart to 1080p and save it to disk before the skin background image loader can access it.

By comparison, in the "fast" video the artwork shown comes from FanartTV:
https://images.fanart.tv/fanart/avengers-endgame-5c979f35948c7.jpg 
That artwork is only *996kb* and is *1080p*. As it is not downsampled, Kodi can load the image directly into the background image loader without waiting for the texture cache to save the downsampled image to disk first.

To make it a fair comparison you need to compare two images of similar size and resolution. Also, you can't have first traversed over a higher resolution image as that will add a bottleneck in Kodi's image processing queue.

**Comment by matke-84** [comment]
@jurialmunkey
Ohhhh yes. Now I see that the pictures are different. But wait. Now I'm checking my settings. I set up to have my pictures pulled from fanart.tv. I also inserted my api key. Why then pull fanart image from the tmdb site which is bigger. That may be the problem then.
EDIT: I'm sorry I didn't download the latest version. Looks like you sorted it out so they pulls fanart pictures from fanart.tv as well.

**Comment by Bequiet19** [comment]
@jurialmunkey It is much better with the new update, there is not too much delay. We are getting closer to the speed of a FEN. :-) I want to show you how fast FEN open lists on an android. This is really amazing. With deleted texture13.db and thumbnails, the list of close to 100 movies opens almost immediately. What has he done here to open so fast? Did you manage to see something specific in the code?
https://files.fm/f/nywrsrj87

**Comment by jurialmunkey** [comment]
> @jurialmunkey It is much better with the new update, there is not too much delay. We are getting closer to the speed of a FEN. :-) I want to show you how fast FEN open lists on an android. This is really amazing. With deleted texture13.db and thumbnails, the list of close to 100 movies opens almost immediately. What has he done here to open so fast? Did you manage to see something specific in the code? https://files.fm/f/nywrsrj87

Kodi's texture cache has no relevance to how fast a plugin directory opens. You need to delete all Fen's caches for a fair comparison. It also has far fewer features than TMDbH and provides less metadata.

**Comment by Bequiet19** [comment]
@jurialmunkey I think I found out where FEN got a little on fast. FEN has three scenarios for showing fanarts. The first two as tmdb helper. Without fanart.tv images, with all fanart.tv images and a third where it combines fanart.tv images (clearart, posters...) and tmdb images (mainly background fanart images). But there is another additional option. Choice of tmdb image resolution (low, medium, high and original). By default it is set to high. I guess there's some way to downsize an image and that's why it's displayed quickly. Because when I set it to the original image, the opening speed is reduced and there is a delay when switching from movie to movie. In any case, here is a video of the first opening of the FEN list, no cache, no thumbnails, everything is clean. This is an option where it combines fanart.tv and tmdb images with high resolution images (not original).
https://files.fm/f/x6svewc8s
Sorry please, I don't want to be boring and ungrateful. Progress is going in the right direction. The speed of tmdb helpers is much improved. I really love this addon and it has been an integral part of my computer for a long time but I would really like to use it on android as well. I just watch more and more movies on TV and tablet than on computer.

**Comment by jurialmunkey** [comment]
You need to define what you mean by slow because you're talking about two entirely separate things that have no bearing on each other - the speed an image loads is entirely separate from the speed the plugin directory loads.

I'll reiterate again - TMDbHelper does NOT download any images and the size of the image has no bearing on fast the plugin directory opens. A plugin directory only provides text metadata and image URLs to Kodi - e.g. just like a link to an image is a separate thing from the image it takes you to https://images.fanart.tv/fanart/avengers-endgame-5c979f35948c7.jpg -- linking to a larger image is NOT going to change how fast this comment loads.

Also I would hardly call that fast loading -- that directory in Fen took 11 seconds for it to open from when you clicked on it.

**Comment by Bequiet19** [comment]
@jurialmunkey Well, I mean both. Open a list and switch from movie to movie. It is important to get a decent speed of both things to get the usability of using addon on android devices.
I don't know how else to explain when the option is set in FEN to the original image and high resolution. When it is set on the original image then everything is slower. Maybe there is another explanation but I don't know how to explain it.
Yes but this is the worst case scenario. FEN opens here a list of almost 100 movies here. A list of 20 movies opens in few seconds.

**Comment by jurialmunkey** [comment]
> I don't know how else to explain when the option is set in FEN to the original image and high resolution.

The fanarttv api doesn't provide options for downscaled images. It is only possible to get downscaled images from TMDb.

The only downscaling option provided by fanarttv is preview which I doubt you want as it is a very small image:
https://images.fanart.tv/preview/avengers-endgame-5c979f35948c7.jpg

By comparison TMDb provides multiple options (there are even lower than these but I think below 780x439 it is getting too small for fanart).

| Quality | Example | Size |
| :--- | :--- | :--- |
| Original | https://www.themoviedb.org/t/p/original/7RyHsO4yDXtBv1zUU3mTpHeQ0d5.jpg | 1607kb |
| 1280x720 | https://www.themoviedb.org/t/p/w1280/7RyHsO4yDXtBv1zUU3mTpHeQ0d5.jpg | 202kb |
| 780x439 | https://www.themoviedb.org/t/p/w780/7RyHsO4yDXtBv1zUU3mTpHeQ0d5.jpg | 84kb |

I can add an option to enable switching to 720p images but it will mean your artwork comes from TMDb not FanartTV.

**Comment by Bequiet19** [comment]
@jurialmunkey Yes, I mentioned that this is an option for tmdb images, not for fanart.tv. Well you are right, I use fanart.tv images in tmdb helper. But if I understood tmdb helper well, where there are no images on fanart.tv, they use as fallback images from the tmdb site. So that option that might help a little.
Also you can add the option to combine fanart.tv and tmdb images. But only as an option, since I like to use all the images from fanart.tv on my computer. They are much better and often the pictures from tmdb site have text on them, so they look like posters and I don't like that. But as an alternative for android it would be ok.

**Comment by jurialmunkey** [comment]
> They are much better and often the pictures from tmdb site have text on them, so they look like posters and I don't like that. 

This is fixable. I can filter these out because TMDb provides data about the language. It's how I determine landscape art (backdrop with iso_639_1 value), so I can just do the reverse for fanart (backdrop without iso_639_1 value).

> Also you can add the option to combine fanart.tv and tmdb images.

I don't see much point in an extra option. Linking the preference order to the quality option should suffice for all scenarios:
e.g. 
FanartTV On + Original Quality = FanartTV first then TMDb as fallback
FanartTV On + Lower Quality = TMDb first then FanartTV as fallback
FanartTV Off = TMDb only.

**Comment by Bequiet19** [comment]
@jurialmunkey This is not a bad idea. But I might put an option for tmdb globally to we had a choice in each scenario for original and lower quality. But I don't know how big job this is.

**Comment by jurialmunkey** [comment]
> @jurialmunkey This is not a bad idea. But I might put an option for tmdb globally to we had a choice in each scenario for original and lower quality. But I don't know how big job this is.

I'm sorry, I don't understand what you mean here. What do you mean by "an option for tmdb globally".

This would already give you an option for quality. There is no point reducing TMDb fanart quality but choosing to have fanart come from fanarttv. Any movie or show that exists on either service will at minimum have fanart and poster so there will never be a fallback for those. The fallbacks only matter for less common artwork types like clearlogo and landscape

**Comment by Bequiet19** [comment]
@jurialmunkey Maybe I didn't understand you. Will there be an option to choose what quality the tmdb image (not fanart.tv) will be. For example, I chose a lower quality tmdb fanart image option and such images will be displayed to me. Did I understand correctly.

**Comment by jurialmunkey** [comment]
> @jurialmunkey Maybe I didn't understand you. Will there be an option to choose what quality the tmdb image (not fanart.tv) will be. For example, I chose a lower quality tmdb fanart image option and such images will be displayed to me. Did I understand correctly.

It is not possible to choose quality for fanarttv. It is only possible to choose quality for TMDb.

**Comment by jurialmunkey** [comment]
^^ v4.5.26 adds an artwork quality option for TMDb under General settings.

**Original**
Uses highest quality available. 
Prefers fanarttv artwork (if enabled and available)

**High**
Uses 1280x720 fanart from TMDb
Prefers fanarttv for all other artwork types (if enabled and available)

**Low**
Uses 780x439 fanart and 342x502 posters from TMDb
Prefers fanarttv for all other artwork types (if enabled and available)

**Comment by Bequiet19** [comment]
@jurialmunkey Woooow. That's it. Now it opens really nicely. :-) Just tell me does tmdb helper use episodes thumnails from fanart.tv or tmdb site? If they use it from the tmdb site, can we include them in this option or maybe it's already included.

**Comment by Bequiet19** [comment]
@jurialmunkey I'm playing with the new setting a bit now. With a low setting, loading and switching from movie to movie is really fast. With high settings are also good performance.
Now the only thing missing is that, if possible at all, we can get litle faster entering in the list. The average entry time is now 7 seconds. If we could get down to 5 seconds it would be perfect. And if that’s not possible, this is a very good improvement over before. Thank you sir for trying to improve this fantastic addon. You're a really good man.

**Comment by Bequiet19** [comment]
> 
> **High** Uses 1280x720 fanart from TMDb Prefers fanarttv for all other artwork types (if enabled and available)
> 
Are you sure they uses everything from fanart.tv and only fanart from tmdb site? I think they pull posters from the tmdb site.

**Comment by jurialmunkey** [comment]
> @jurialmunkey Woooow. That's it. Now it opens really nicely. :-) Just tell me does tmdb helper use episodes thumnails from fanart.tv or tmdb site? If they use it from the tmdb site, can we include them in this option or maybe it's already included.

They come from TMDb (fanarttv doesn't have thumbs). Your comment made me realise I missed those though! Should be fixed now.

**Comment by Bequiet19** [comment]
@jurialmunkey Thanks. Loading episodes is faster as well as switching between episodes.
What do you think, does entering in lists can be accelerated just a little more? Also entry from the tv show into the episodes also opens litle slowly. If that could be accelerated a bit, it would be great.

**Comment by matke-84** [comment]
@jurialmunkey 
> This is fixable. I can filter these out because TMDb provides data about the language. It's how I determine landscape art (backdrop with iso_639_1 value), so I can just do the reverse for fanart (backdrop without iso_639_1 value).

This is a good idea for a fanart from TMDb. Some of them have a bunch of captions. For fanart it doesn't matter language.

**Comment by xyzfre** [comment]
Great work on this new update I use a Nvidia shield on android 11 which sucks lol but I do see a difference on fanart images thank you for all the hard work that goes into TMDb Helper truly is an amazing project

**Comment by xyzfre** [comment]
I use AuraMod and I just noticed a few shows that were on my next episodes we're missing episode thumbs I never made an issue about it because a couple were very old shows or those weird ones that nobody watches lol and now they all show up properly GREAT JOB👍🏼

**Comment by matke-84** [comment]
@jurialmunkey 
I noticed that some of the original logos are very large. Because of that some movie logos open with delay. What do you think we should do a similar thing with logo like with posters and fanart, except that I would only go with size W500. Everything below that is little for skins.
EDIT: Actually can also w300 for Low setting. It doesn't look so bad.

**Comment by jurialmunkey** [comment]
> @jurialmunkey I noticed that some of the original logos are very large. Because of that some movie logos open with delay. What do you think we should do a similar thing with logo like with posters and fanart, except that I would only go with size W500. Everything below that is little for skins. EDIT: Actually can also w300 for Low setting. It doesn't look so bad.

Definitely, was thinking something similar as I was noticing the resolution on some of the logos from TMDb is massive - way bigger than necessary. 

Added. I set it to w500 for high/medium and w300 for low. I was going to reduce it on original too but changed my mind. Ideally there would be a w800 setting to bring it in line with fanarttv (ftv has 800 and 400 for clearlogos) - but original will have to do.

**Comment by matke-84** [comment]
@jurialmunkey
You're reading my mind. I just wanted to write you the same thing (for original it is better be a bigger logo), but like I said, you're fast on the trigger. :-)

**Comment by matke-84** [comment]
> @jurialmunkey
> 
> > This is fixable. I can filter these out because TMDb provides data about the language. It's how I determine landscape art (backdrop with iso_639_1 value), so I can just do the reverse for fanart (backdrop without iso_639_1 value).
> 
> This is a good idea for a fanart from TMDb. Some of them have a bunch of captions. For fanart it doesn't matter language.

Don't forget this.

**Comment by jurialmunkey** [comment]
> > @jurialmunkey
> > > This is fixable. I can filter these out because TMDb provides data about the language. It's how I determine landscape art (backdrop with iso_639_1 value), so I can just do the reverse for fanart (backdrop without iso_639_1 value).
> > 
> > 
> > This is a good idea for a fanart from TMDb. Some of them have a bunch of captions. For fanart it doesn't matter language.
> 
> Don't forget this.

Should be done now.  Will only work with extra details from TMDb enabled. Might need to refresh details of item if it is already cached.

**Comment by matke-84** [comment]
> Should be done now. Will only work with extra details from TMDb enabled. Might need to refresh details of item if it is already cached.

Yes. It works like a charm. Now we have almost perfect fanarts. Now I'm testing on android. And you know what? This all works very well. 😀 The only thing left is to try to speed up the entry in the lists (movies, tv shows, seasons...) on android devices. It's a little slower but this is such a bust compared to the old versions. I know this is going to be a little harder since there is a lot of metadata pulled by tmdb helper but we could think in the direction of making the same option like for fanarts, I don't know, I'm just thinking out loud.

EDIT: My mistake. There was something left in the background so it opened slowly. It actually opens the lists pretty ok. I measured an average speed is 5-6 seconds (real time seems to be two seconds longer) on android. Maybe another two or three seconds of acceleration and it would be perfect.

**Comment by jurialmunkey** [comment]
I need logs with timer reports to improve times further on Android - and ideally in lot of different scenarios with different settings. I don't own any android devices so it is impossible for me to tell what specific areas need optimisation and the only way I can find that out is from timer report logs. 

Without more information I'm just stumbling around in the dark guessing at what might improve things.

**Comment by Bequiet19** [comment]
Wooow. Great improvement. Now it seems to open even faster. A little more speed and it will be fantastic.

**Comment by matke-84** [comment]
@jurialmunkey 
OK. I managed to extract the logs from the android. I tried as many different scenarios as possible. I have to say I installed kodi from scratch. I just installed one skin and tmdb helper that is conected with my trakt account. I don't have a lot of movies and tv shows on my trakt account so I don't know if it would slow down even more if I withdrew that information as well. We can say that this is an ideal scenario, I don't know how it would behave with multiple addons and skins installed. These are the results but it seems to me that in practice it opens a little slower, about two seconds is the difference if I measure with a stopwatch. The results are different, but some average is about 5 seconds per log, or about 6-7 seconds in reality. Some lists open slower and also widgets (although this is not so relevant, I just put it as an example). The log also includes the opening of seasons and episodes.
If you need anything else, I'm here.
https://paste.kodi.tv/devogixovu

**Comment by jurialmunkey** [comment]
Thanks @matke-84 that's exactly what I was looking for. There's a couple of outliers where the unusually longer times can be probably be ignored and blamed on external factors, but otherwise looks like some consistent trends about which areas need improvement and which don't. 

Interesting to see TMDb details costing more than Fanarttv on Android now, as even with the recent improvements it's still generally the other way around for me on pc.

I've got a couple of ideas but might take a while to implement as it's a bigger change.

**Comment by matke-84** [comment]
@jurialmunkey
These are results with various settings. What is interesting when it enters in the list, everything now looks great and fast, but the process of entering in the list, seasons, episodes takes some time. I even noticed that even changing the quality settings does not change the situation of entering in the list too much.

**Comment by jurialmunkey** [comment]
> @jurialmunkey These are results with various settings. What is interesting when it enters in the list, everything now looks great and fast, but the process of entering in the list, seasons, episodes takes some time. I even noticed that even changing the quality settings does not change the situation of entering in the list too much.

Yeah I've know for a while that seasons/episodes needs optimising. The current approach isn't particularly great in terms of efficiency for joining the parent tvshow details. But that's something I'll deal with once I get a little more performance out of the standard lists.

**Comment by jurialmunkey** [comment]
> ... the process of entering in the list, seasons, episodes takes some time.

@matke-84 - On the latest version you should see a decent improvement on season/episode list loading - at least I definitely am seeing improvements now.

**Comment by matke-84** [comment]
@jurialmunkey
I'll take measurements tomorrow and we'll see the results. Unfortunately, at first view, it doesn't seem to me that there is improvement.

**Comment by matke-84** [comment]
@jurialmunkey
There are litle improvements. After completely deleting cache season opens pretty ok while episodes i think same as before. I'll do a complete analysis tomorrow.

**Comment by Bequiet19** [comment]
@jurialmunkey Unfortunately, I did not notice any improvement in opening speed. But that's why every switch from movie to movie works so nice.

**Comment by jurialmunkey** [comment]
> @jurialmunkey Unfortunately, I did not notice any improvement in opening speed. But that's why every switch from movie to movie works so nice.

I wasn't talking about movies. I was specifically talking about going into the seasons or episodes list of a tvshow.

**Comment by Bequiet19** [comment]
@jurialmunkey Yes, I'm sorry sir. I’m also talking about opening seasons and episodes. The opening of seasons and episodes is still late. Average about 5-6 seconds to open the season. Do I need to do something to speed up?

**Comment by jurialmunkey** [comment]
> @jurialmunkey Yes, I'm sorry sir. I’m also talking about opening seasons and episodes. The opening of seasons and episodes is still late. Average about 5-6 seconds to open the season. Do I need to do something to speed up?

v4.5.40/41 had a bug where fanarttv lookups were being duplicated for episodes so you wouldn't have seen the benefit if you were using that version. Fixed in v4.5.42 so you should see a decent improvement opening episodes using that version.

I've also delayed writing disk cache until after directory is created in the hope that might improve performance on devices with slow disk write speeds.

**Comment by Bequiet19** [comment]
@jurialmunkey I don't know, I don't see any acceleration. Maybe I'm doing something wrong. It seems to me a bit slower with the latest version.

**Comment by jurialmunkey** [comment]
> @jurialmunkey I don't know, I don't see any acceleration. Maybe I'm doing something wrong. It seems to me a bit slower with the latest version.

I had to revert some changes due to an unexpected issue. https://github.com/jurialmunkey/plugin.video.themoviedb.helper/issues/640#issuecomment-1037078391

**Comment by Bequiet19** [comment]
@jurialmunkey
In the latest version, there are a some improvements in opening seasons and episodes. Is there any chance for movie and TV show lists to do the same?

**Comment by matke-84** [comment]
@jurialmunkey
I was thinking something. If we increase the number of connections to all services, would we get some speed improvement?

**Comment by jurialmunkey** [comment]
> @jurialmunkey I was thinking something. If we increase the number of connections to all services, would we get some speed improvement?

What do you mean by increase the number of connections? Each item in the list is already given it's own thread for API requests.

**Comment by matke-84** [comment]
@jurialmunkey
I thought that if increased number of connections , that can we number of incoming/outgoing requests could also increase, but I didn't know that it was determined by api. Sorry my bad.

**Comment by jurialmunkey** [comment]
> @jurialmunkey I thought that if increased number of connections , that can we number of incoming/outgoing requests could also increase, but I didn't know that it was determined by api. Sorry my bad.

Every request happens as its own thread with its own connection.

For instance, see this timer report

```
DIRECTORY TIMER REPORT
info=popular&page=6&plugin_category=Popular%20Movies&tmdb_id=None&tmdb_type=movie
------------------------------
get_list       :   0.455 sec
 - kodi_db     :   0.007 sec
add_items      :   1.825 sec
--build        :   1.742 sec
 - item_api    :   1.644 sec avg |   1.723 sec max |  20
 - item_ftv    :   1.615 sec avg |   1.699 sec max |  20
 - item_tmdb   :   0.549 sec avg |   0.779 sec max |  19
--make         :   0.075 sec
------------------------------
Total          :   2.280 sec
```

To create this list, there were 20 requests to FanartTV taking an average of `1.615s` per request, and 19 requests to TMDb taking an average of `0.549s` per request. To do these 39 requests one at a time would take `42.731s`. However, we can see from the `--build` time that due to threading it only took `1.742s` _total_ to get ALL 39 of those requests AND also do the computation to build all 20 items.

That total time is only `0.043s` longer than the longest FanartTV request of `1.699s`. That means even if all the other items were already cached and we only needed to do that one single request to FanartTV, it would still take `1.699s` whilst we waited for FanartTV to send the data. 

There is practically no optimisation available here - we cannot go faster than it takes the FanartTV server to send the data.

From the above, we can see that the only _non_ api based cost involved was `0.075s` finalising the items and sending to Kodi plugin directory (`--make`), the `0.007s` getting the Kodi db from cache, and the `0.043s` remainder in the build section. That's a grand total of `0.125s` doing pure computation tasks that don't involve waiting for APIs. That is a trivial number.

**Comment by matke-84** [comment]
@jurialmunkey
With the latest update landscape season specific artworks don't work. Instead of a landscape image of the season, it shows me a landscape image of the whole tv show.

**Comment by matke-84** [comment]
> > @jurialmunkey I thought that if increased number of connections , that can we number of incoming/outgoing requests could also increase, but I didn't know that it was determined by api. Sorry my bad.
> 
> Every request happens as its own thread with its own connection.
> 
> For instance, see this timer report
> 
> ```
> DIRECTORY TIMER REPORT
> info=popular&page=6&plugin_category=Popular%20Movies&tmdb_id=None&tmdb_type=movie
> ------------------------------
> get_list       :   0.455 sec
>  - kodi_db     :   0.007 sec
> add_items      :   1.825 sec
> --build        :   1.742 sec
>  - item_api    :   1.644 sec avg |   1.723 sec max |  20
>  - item_ftv    :   1.615 sec avg |   1.699 sec max |  20
>  - item_tmdb   :   0.549 sec avg |   0.779 sec max |  19
> --make         :   0.075 sec
> ------------------------------
> Total          :   2.280 sec
> ```
> 
> To create this list, there were 20 requests to FanartTV taking an average of `1.615s` per request, and 19 requests to TMDb taking an average of `0.549s` per request. To do these 39 requests one at a time would take `42.731s`. However, we can see from the `--build` time that due to threading it only took `1.742s` _total_ to get ALL 39 of those requests AND also do the computation to build all 20 items.
> 
> That total time is only `0.043s` longer than the longest FanartTV request of `1.699s`. That means even if all the other items were already cached and we only needed to do that one single request to FanartTV, it would still take `1.699s` whilst we waited for FanartTV to send the data.
> 
> There is practically no optimisation available here - we cannot go faster than it takes the FanartTV server to send the data.
> 
> From the above, we can see that the only _non_ api based cost involved was `0.075s` finalising the items and sending to Kodi plugin directory (`--make`), the `0.007s` getting the Kodi db from cache, and the `0.043s` remainder in the build section. That's a grand total of `0.125s` doing pure computation tasks that don't involve waiting for APIs. That is a trivial number.

I see. Truly a negligible difference.

**Comment by jurialmunkey** [comment]
> @jurialmunkey With the latest update landscape season specific artworks don't work. Instead of a landscape image of the season, it shows me a landscape image of the whole tv show.

Should be fixed now.

**Comment by matke-84** [comment]
@jurialmunkey
It works. 👍
I tried this addon that mention @Bequiet19, Fen, I've never heard of it before but it opens lists really fast in a fact. I guess it pulls less metadata but I didn't notice that something was missing also artworks are also included. But what is not clear to me, how managed to open seasons and episodes instantly, without a second of waiting. Does it pull this metadata and put it somewhere in cache when opening the lists, but in other hand it is not possible for me because open list very quickly. I know that tmdb helper and this addon are not comparable but maybe we can find some solution here as well.

**Comment by jurialmunkey** [comment]
> I guess it pulls less metadata but I didn't notice that something was missing also artworks are also included. But what is not clear to me, how managed to open seasons and episodes instantly, without a second of waiting. Does it pull this metadata and put it somewhere in cache when opening the lists, but in other hand it is not possible for me because open list very quickly. I know that tmdb helper and this addon are not comparable but maybe we can find some solution here as well.

It's not a fair comparison. It's like comparing a motorbike with a truck and asking if there's something the motorbike is doing to make it go faster than the truck. 

Fen's fanarttv implementation is fast because it is incredibly simplistic. And that's because Fen only needs the data for one thing: providing a list of items. Plus, it doesn't look up season specific artwork or allow manual selection.

And on the metadata front, you might not notice it as an end user but you are only using a very small subset of the features that TMDbH provides. TMDbH's first role is a skinning tool, so it gives far richer metadata. See the detailed item wiki for examples https://github.com/jurialmunkey/plugin.video.themoviedb.helper/wiki/Detailed-Item 

As a simple example, I guaranteed Fen cannot give you `ListItem.Property(sound_department.1.name)` or `ListItem.Property(photography.1.name)`. The rich metadata in TMDbH is important for skinners who want to do complex filtering to create related lists - for instance, say I want a list of all other movies with the same director of photography in the same genre - TMDbH provides tools for skinners to do that.

**Comment by matke-84** [comment]
@jurialmunkey
I said that these two addons are not comparable.
I didn't even compare them, tmdb helper is most complex addon, with much more metadata and functionality.
But let’s say elementum that’s done in go, which is very fast, and I believe it’s nothing much more complex than Fen, doesn’t open seasons and episodes instantly like a Fen. That was weird to me. So I was wondering if there might be in code something specifically that what we can use, nothing else.

**Comment by matke-84** [comment]
@jurialmunkey
With the latest update, we lost the ability to choose a fanart background from fanart.tv in artwork manager.

**Comment by Bequiet19** [comment]
@jurialmunkey It seems to me that the lists open a little slower again compared to the previous version I used (I think version 4.5.68). Now it opens the lists again in about 6-7 seconds. Only if we could go down to 3-4 seconds would it be perfect. I really hope so sir. Everything else is great. Fast switching between movies, artwoks...

**Comment by jurialmunkey** [comment]
> @jurialmunkey It seems to me that the lists open a little slower again compared to the previous version I used (I think version 4.5.68). Now it opens the lists again in about 6-7 seconds. Only if we could go down to 3-4 seconds would it be perfect. I really hope so sir. Everything else is great. Fast switching between movies, artwoks...

Please provide timer logs otherwise I cannot help you. I need the timer log data to be able to identify the issue

**Comment by xyzfre** [comment]
I'm sorry not to be disrespectful but I don't think two seconds make a big difference TMDb runs great to mess something up just to increase two or three seconds on my end it's not worth it,, TMDb works amazingly well already thank you for all the hard work that goes into this project

**Comment by Bequiet19** [comment]
@

> > @jurialmunkey It seems to me that the lists open a little slower again compared to the previous version I used (I think version 4.5.68). Now it opens the lists again in about 6-7 seconds. Only if we could go down to 3-4 seconds would it be perfect. I really hope so sir. Everything else is great. Fast switching between movies, artwoks...
> 
> Please provide timer logs otherwise I cannot help you. I need the timer log data to be able to identify the issue

```
DIRECTORY TIMER REPORT
                                                   info=trakt_popular&tmdb_type=movie
                                                   ------------------------------
                                                   idx_lookup     :   0.018 sec
                                                   get_list       :   1.692 sec
                                                    - kodi_db     :   None
                                                   add_items      :   4.491 sec
                                                   --build        :   3.961 sec
                                                    - item_api    :   2.991 sec avg |   3.695 sec max |  20
                                                    - item_ftv    :   1.651 sec avg |   2.577 sec max |  20
                                                    - item_tmdb   :   2.678 sec avg |   3.489 sec max |  20
                                                   --make         :   0.467 sec
                                                   ------------------------------
                                                   Total          :   6.185 sec

                                                   item_api:
                                                   1.936  2.194  2.382  3.018  3.046  2.922  3.059  3.061  3.177  3.184  3.194  2.564  3.154  3.608  3.695  3.053  3.031  3.624  2.908  3.006 
                                                   
                                                   item_ftv:
                                                   1.082  1.121  1.176  1.231  1.289  1.385  1.396  1.305  1.508  0.976  1.636  1.485  2.108  1.909  2.474  1.803  2.259  2.192  2.116  2.577 
                                                   
                                                   item_tmdb:
                                                   1.749  1.952  2.186  2.447  2.595  2.633  2.768  2.602  2.856  3.004  3.037  2.349  2.531  3.162  3.489  2.803  2.743  2.564  3.405  2.677 
```
This is average. Sometimes it's more.

**Comment by jurialmunkey** [comment]
@Bequiet19 - That's just slow internet connection and/or slow server response times.

In the `get_list` line you can see it took `1.692 sec` to receive the list data back from the TMDb server. And that's before we can even start looking up item details.

In the `item_tmdb` line you can see that the longest item details request took `3.489 sec` for the data to come back from TMDb. That means at very minimum, we would need to wait `5.181 sec` for the data (time for list + longest item).

The `total` line shows that the **entire** directory only took `6.185s`. That's a mere `1.004s` on top of the slowest request to do **40** additional requests (1xTrakt, 20xFanartTV, plus the 19x other item details from TMDb) plus process the data.

This is why the timer reports are important. I can't make the internet go faster. If the server is slow, the server is slow.

**Comment by Bequiet19** [comment]
@jurialmunkey
I read that tmdb helper has more data that it downloads from the server and cannot be compared to any other addon which open lists almost immediately because of their simplicity. Is there a possibility to make an option that would give the end user the opportunity to choose what he needs, not to burden the Internet or the device with unnecessary data. Then I guess we would gain in speeding up opening lists? Sorry if I ask stupid questions, I don't understand much about python programming. I would just like to use tmdb helper on android with similar experience as on computer because this is one of my favorite addons. :-)

**Comment by matke-84** [comment]
> @jurialmunkey I read that tmdb helper has more data that it downloads from the server and cannot be compared to any other addon which open lists almost immediately because of their simplicity. Is there a possibility to make an option that would give the end user the opportunity to choose what he needs, not to burden the Internet or the device with unnecessary data. Then I guess we would gain in speeding up opening lists? Sorry if I ask stupid questions, I don't understand much about python programming. I would just like to use tmdb helper on android with similar experience as on computer because this is one of my favorite addons. :-)

I'm not really sure if this is possible but you will get an answer from @jurialmunkey. What I want to advise you is to try to delete all cache (tmdb helper and kodi), thumbs... and then reset the device. Android devices know that they sometimes slow down due to accumulated cache and lack of space.

**Comment by jurialmunkey** [comment]
> @jurialmunkey I read that tmdb helper has more data that it downloads from the server and cannot be compared to any other addon which open lists almost immediately because of their simplicity. Is there a possibility to make an option that would give the end user the opportunity to choose what he needs, not to burden the Internet or the device with unnecessary data. Then I guess we would gain in speeding up opening lists? Sorry if I ask stupid questions, I don't understand much about python programming. I would just like to use tmdb helper on android with similar experience as on computer because this is one of my favorite addons. :-)

There are already options to turn off FanartTV, TMDb extra details, and Trakt watched data. I have told you that right from the very start.

Regardless, the requests size will only be a few kilobytes because it is compressed plain text. For instance, the full data for `Alien (1979)` is `~13kb` (13298 bytes to be exact) of compressed data (it decompresses to `94kb`). Even on ADSL that'd be under 0.02s to download. The data size is trivial - reducing it from 13Kb to 10Kb isn't noticeable.

The time impact is server response time (basically ping speed). The server can only server so many users at a time. If there's high demand, the wait time in the queue to ask for the data is longer (just like being on hold on the phone). If the queue gets long then TMDb starts rate limiting users. Once you hit the rate limit you get pushed back in the queue.

The `get_list` stage gives a good indication of how long TMDb is taking to respond because there's not much else TMDbHelper can do until it gets the lists data. A normal time is between 0.4 to 0.8s. That you're getting 1.6s indicates high server load (or a bad internet connection or other services locking the cpu) causing poor times rather than TMDbHelper.


To demonstrate: See below for two pages from the same list on the same machine with cache deleted. The machine is an i7 with SSD and 32gb ram, so we know we aren't being impacted by poor CPU performance. You can see the first one I open the page in 1.8s whereas the second blew out to almost 5s.

On the first, my get_list time was only `0.410s` and my request max never went over `1.400s`. These are normal response times. As a result, the list was fast to open with `1.871s`

In the second, you can see my get_list time was more than double at `0.894s` and one of my TMDb requests maxed out at `3.730s`. These are much slower response times than normal. Clearly there was some network/server congestion  and as a result, that particular page took _way_ longer to load -- almost 5 seconds.


```
DIRECTORY TIMER REPORT
info=popular&page=6&plugin_category=Popular%20Movies&tmdb_id=None&tmdb_type=movie
------------------------------
idx_lookup     :   0.002 sec
get_list       :   0.410 sec
 - kodi_db     :   0.008 sec
add_items      :   1.460 sec
--build        :   1.369 sec
 - item_api    :   1.010 sec avg |   1.368 sec max |  20
 - item_ftv    :   0.983 sec avg |   1.079 sec max |  20
 - item_tmdb   :   0.635 sec avg |   1.364 sec max |  20
--make         :   0.086 sec
------------------------------
Total          :   1.871 sec
```

```
DIRECTORY TIMER REPORT
info=popular&page=3&plugin_category=Popular%20Movies&tmdb_id=None&tmdb_type=movie
------------------------------
idx_lookup     :   0.003 sec
get_list       :   0.894 sec
 - kodi_db     :   0.009 sec
add_items      :   4.049 sec
--build        :   3.836 sec
 - item_api    :   1.150 sec avg |   3.751 sec max |  20
 - item_ftv    :   1.005 sec avg |   1.091 sec max |  20
 - item_tmdb   :   0.795 sec avg |   3.730 sec max |  17
--make         :   0.205 sec
------------------------------
Total          :   4.944 sec
```

**Comment by matke-84** [comment]
There has been a big drop in performance. It opens lists very slowly. Here's the log.

------------------------------
                                                   get_list       :   0.252 sec
                                                   add_items      :   8.933 sec
                                                   --build        :   8.554 sec
                                                    - item_api    :   6.898 sec avg |   8.531 sec max |  19
                                                    - item_ftv    :   7.446 sec avg |   7.937 sec max |  16
                                                    - item_tmdb   :   7.737 sec avg |   8.200 sec max |  16
                                                   --make         :   0.344 sec
                                                   ------------------------------
                                                   Total          :   9.186 sec
                                                   
                                                   item_api:
                                                   0.057  0.093  0.092  7.696  7.837  8.024  8.051  8.031  8.070  8.078  8.112  8.116  8.074  8.330  8.489  8.483  8.531  8.384  8.509 
                                                   
                                                   item_ftv:
                                                   6.885  7.061  7.040  7.107  7.183  7.203  7.285  7.367  7.553  7.633  7.681  7.648  7.749  7.887  7.920  7.937 
                                                   
                                                   item_tmdb:
                                                   7.091  7.212  7.377  7.445  7.526  7.534  7.708  7.713  7.758  7.806  7.941  8.141  8.064  8.155  8.115  8.200

------------------------------
                                                   get_list       :   0.230 sec
                                                   add_items      :   6.329 sec
                                                   --build        :   5.579 sec
                                                    - item_api    :   4.405 sec avg |   5.566 sec max |  19
                                                    - item_ftv    :   4.435 sec avg |   5.166 sec max |  16
                                                    - item_tmdb   :   4.784 sec avg |   5.351 sec max |  16
                                                   --make         :   0.709 sec
                                                   ------------------------------
                                                   Total          :   6.560 sec
                                                   
                                                   item_api:
                                                   0.104  0.110  0.112  4.434  4.756  5.001  5.064  5.077  5.121  5.141  5.147  5.230  5.319  5.390  5.492  5.529  5.537  5.566  5.565 
                                                   
                                                   item_ftv:
                                                   3.941  3.893  3.932  4.084  4.075  4.216  4.255  4.320  4.464  4.538  4.666  4.644  4.896  4.927  4.950  5.166 
                                                   
                                                   item_tmdb:
                                                   3.750  4.083  4.423  4.478  4.558  4.700  4.725  4.568  4.882  4.892  5.085  5.195  5.250  5.351  5.296  5.314

------------------------------
                                                   get_list       :   0.155 sec
                                                   add_items      :  14.300 sec
                                                   --build        :  13.726 sec
                                                    - item_api    :  13.130 sec avg |  13.715 sec max |  20
                                                    - item_ftv    :  10.408 sec avg |  13.241 sec max |  24
                                                    - item_tmdb   :  12.611 sec avg |  13.463 sec max |  20
                                                   --make         :   0.547 sec
                                                   ------------------------------
                                                   Total          :  14.457 sec
                                                   
                                                   item_api:
                                                   12.217  12.255  12.854  12.887  12.908  12.968  12.982  13.055  13.066  13.102  13.172  13.247  13.274  13.309  13.331  13.416  13.506  13.633  13.710  13.715 
                                                   
                                                   item_ftv:
                                                   11.220  11.540  11.599  11.961  11.882  12.045  12.143  12.378  12.417  0.110  12.718  12.651  12.708  0.139  0.069  0.059  12.839  12.911  12.989  12.906  13.022  13.109  13.125  13.241 
                                                   
                                                   item_tmdb:
                                                   11.777  11.779  11.991  12.029  12.078  12.162  12.425  12.433  12.496  12.438  12.539  12.683  12.873  12.857  13.050  13.165  13.186  13.329  13.463  13.461

------------------------------
                                                   get_list       :   0.111 sec
                                                   add_items      :   9.623 sec
                                                   --build        :   9.215 sec
                                                    - item_api    :   8.393 sec avg |   9.206 sec max |  20
                                                    - item_ftv    :   7.756 sec avg |   8.974 sec max |  20
                                                    - item_tmdb   :   8.527 sec avg |   9.054 sec max |  19
                                                   --make         :   0.373 sec
                                                   ------------------------------
                                                   Total          :   9.736 sec
                                                   
                                                   item_api:
                                                   0.060  8.154  8.203  8.191  8.540  8.563  8.611  8.684  8.804  8.868  9.038  9.074  9.056  9.077  9.076  9.089  9.192  9.169  9.200  9.206 
                                                   
                                                   item_ftv:
                                                   7.326  7.455  7.562  7.591  7.609  7.551  7.871  7.973  8.078  8.072  8.347  8.358  8.475  0.062  8.702  8.731  8.683  8.800  8.907  8.974 
                                                   
                                                   item_tmdb:
                                                   7.921  7.947  8.068  8.182  8.072  8.252  8.433  8.487  8.411  8.527  8.466  8.810  8.714  8.854  8.904  8.876  8.984  9.054  9.045

------------------------------
                                                   get_list       :   0.953 sec
                                                   add_items      :  18.800 sec
                                                   --build        :  18.336 sec
                                                    - item_api    :  16.679 sec avg |  18.295 sec max |  20
                                                    - item_ftv    :  15.000 sec avg |  16.583 sec max |  20
                                                    - item_tmdb   :  16.095 sec avg |  16.707 sec max |  19
                                                   --make         :   0.391 sec
                                                   ------------------------------
                                                   Total          :  19.805 sec
                                                   
                                                   item_api:
                                                   0.106  17.118  16.787  17.502  17.333  17.253  16.992  17.901  17.373  17.782  17.208  17.694  17.464  17.445  17.994  18.209  17.476  18.295  17.679  17.970 
                                                   
                                                   item_ftv:
                                                   15.210  15.226  15.047  15.402  15.374  15.366  15.512  15.584  15.731  15.572  16.160  16.181  16.190  16.017  15.988  16.108  16.129  16.583  0.116  16.502 
                                                   
                                                   item_tmdb:
                                                   15.331  15.665  15.424  15.726  15.616  15.879  15.912  15.912  15.935  16.279  16.650  16.178  16.593  16.171  16.209  16.478  16.689  16.449  16.707

**Comment by matke-84** [comment]
@jurialmunkey Now everything is ok again. I don't know what happened. I tried to clear cache, reset kodi... and suddenly everything went back to normal.

**Comment by jurialmunkey** [comment]
> @jurialmunkey Now everything is ok again. I don't know what happened. I tried to clear cache, reset kodi... and suddenly everything went back to normal.

Looks like you had just had a slow internet/network connection for a while. Nothing has changed on the item builder front. 

"item-ftv" _only_ times how long it takes to check the cache and do a look from fanarttv if the item doesn't exist. So a slow time in that section pretty much can only mean a slow internet connection (if it was a disk issue you'd get an error about the database being locked).

**Comment by matke-84** [comment]
@jurialmunkey
It's not up to internet speed. It is up to the skin helper service, more precisely metadatautils module. After the last change tmdb helper slowed down a lot. Something about multiprocessing and android has been deleted.
https://github.com/kodi-community-addons/script.module.metadatautils/commit/00c3af7bf48d7f5062598c9a1cae1ead9cfe3c8d
When I return the version without this change, everything works normally.

**Comment by jurialmunkey** [comment]
@matke-84 - I don't use metadatautils module and  I don't use multiprocessing. Neither of those things have anything to do with TMDbH.

If some other addon is messing with your internet connection speed then I suggest you remove it.

**Comment by matke-84** [comment]
@jurialmunkey
You didn't understand me. I didn't say that there was anything with tmdb helper, I said that it was to metadatautils, and I have to use it because it is necessary for my skin. I just mentioned it kills performance, among other things, and tmdb helpers. That's why the slowdown happened. This has nothing to do with you and tmdb helper. ;-)

**Comment by jurialmunkey** [comment]
> @jurialmunkey You didn't understand me. I didn't say that there was anything with tmdb helper, I said that it was to metadatautils, and I have to use it because it is necessary for my skin. I just mentioned it kills performance, among other things, and tmdb helpers. That's why the slowdown happened. This has nothing to do with you and tmdb helper. ;-)

Ah, I see what you're saying! Interesting that it would cause such a performance hit to other add-ons not using it - it must be completely maxing out the cpu.
